urllib3/util/connection.py

```diff
@@ -108,6 +116,134 @@ def _set_socket_options(sock, options):
         sock.setsockopt(*opt)
 
 
+def _connect_happy_eyes(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
+                        source_address=None, socket_options=None):
+    """ Implements the Happy Eyes protocol (RFC 6555) which allows
+    multiple sockets to attempt to connect from different families
+    for better connect times for dual-stack clients where server
+    IPv6 service is advertised but broken. """
+    err = None
+
+    if address in HAPPY_EYES_CACHE:
+        family, proto, expires = HAPPY_EYES_CACHE[address]
+
+        # If the cache entry is expired, don't use it.
+        if time.time() < expires:
+            del HAPPY_EYES_CACHE[address]
+
+        # Otherwise try to use the entry right away, if this
+        # fails then run the Happy Eyes protocol again.
+        else:
+            sock = None
+            try:
+                sock = socket.socket(family, socket.SOCK_STREAM, proto)
+
+                _set_socket_options(sock, socket_options)
+
+                if timeout is not socket._GLOBAL_DEFAULT_TIMEOUT:
+                    sock.settimeout(timeout)
+                if source_address:
+                    sock.bind(source_address)
+
+                sock.connect(address)
+                return sock
+            except socket.error as e:
+                err = e
+                sock.close()
+
+    # Couldn't find a cached value or cached value didn't work.
+    host, port = address
+    socks = []
+
+    # Perform a DNS lookup for the address for IPv4 or IPv6 families.
+    family = allowed_gai_family()
+    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
+        af, socktype, proto, canonname, sa = res
+
+        # Create and connect the socket for each address we get back.
+        sock = None
+        try:
+            sock = socket.socket(af, socktype, proto)
+
+            _set_socket_options(sock, socket_options)
+
+            if source_address:
+                sock.bind(source_address)
+            sock.settimeout(0)  # Make the socket non-blocking to use select.
+
+            # We should catch all errors except all non-blocking error codes.
+            errcode = sock.connect_ex(sa)
+            if errcode not in [errno.EINPROGRESS,
+                               errno.EAGAIN,
+                               errno.ENOTCONN]:
+                raise socket.error(errcode)
+
+            socks.append(sock)
```

Comment:
We may need more complex logic than this.

RFC 6555 says a few things of note. The first thing it says is in Section 5.4, where it points out that:

> It is possible that a DNS query for an A or AAAA resource record will return more than one A or AAAA address.  When this occurs, it is RECOMMENDED that a Happy Eyeballs implementation order the responses following the host's address preference policy and then try the first address.  If that fails after a certain time (see Section 5.5), the next address SHOULD be the IPv4 address.
>
> If that fails to connect after a certain time (see Section 5.5), a Happy Eyeballs implementation SHOULD try the other addresses returned; the order of these connection attempts is not important.

In Section 5.5, it says:

> It is RECOMMENDED that connection attempts be paced to give connections a chance to complete.  It is RECOMMENDED that connection attempts be paced 150-250 ms apart to balance human factors against network load.

So what we may want to do is to take this result and iterate over it, opening a connection, adding it to the selector pool, and then waiting for select to return with a timeout of 200ms before we create the next connection. We wouldn't remove the other connection from the pool, of course: the pool would just get slowly bigger until we finally connect.

This would require some careful math to make sure we don't exceed the total connection timeout, but I think it'd provide a much better behaviour.

Are you up for implementing this?